File: src/com/waverunnah/swg/harvesterdroid/data/resources/GalaxyResource.java
/*
 * HarvesterDroid - A Resource Tracker for Star Wars Galaxies
 * Copyright (C) 2017  Waverunner
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.waverunnah.swg.harvesterdroid.data.resources;

import com.waverunnah.swg.harvesterdroid.app.Attributes;

import javax.persistence.CascadeType;
import javax.persistence.CollectionTable;
import javax.persistence.Column;
import javax.persistence.ElementCollection;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.MapKeyColumn;
import javax.persistence.Table;
import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlElementWrapper;
import javax.xml.bind.annotation.XmlRootElement;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@XmlRootElement(name="galaxy_resource") @XmlAccessorType(XmlAccessType.FIELD)
@Entity @Table(name="GalaxyResource")
public class GalaxyResource {
    @Id @Column(name="Name") @XmlAttribute(name="name")
    private String name;
    @Column(name="SpawnDate") @XmlAttribute(name="spawn_date")
    private Date date;
    @Column(name="Container") @XmlAttribute(name="container")
    private String container;
    @Column(name="DespawnDate") @XmlAttribute(name="despawn_date")
    private Date despawnDate;
    @ElementCollection @CollectionTable(name="GalaxyResource_Planets") @Column(name="Planet")
    @XmlElementWrapper(name="planets") @XmlElement(name="planet")
    private List<String> planets;
    @ElementCollection @MapKeyColumn(name="Attribute") @Column(name="Value") @XmlElementWrapper(name="attributes")
    private Map<String, Integer> attributes;

    @ManyToOne(fetch = FetchType.EAGER, cascade = CascadeType.ALL) @XmlElement(name="resource_type")
    private ResourceType resourceType;

    private transient String resourceTypeString;

    public GalaxyResource() {
        planets = new ArrayList<>();
        attributes = new HashMap<>(Attributes.size());
        Attributes.forEach((primary, secondary) -> attributes.put(primary, -1));
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Date getDate() {
        return date;
    }

    public void setDate(Date date) {
        this.date = date;
    }

    public ResourceType getResourceType() {
        return resourceType;
    }

    public void setResourceType(ResourceType resourceType) {
        this.resourceType = resourceType;
    }

    public String getContainer() {
        return container;
    }

    public void setContainer(String container) {
        this.container = container;
    }

    public Date getDespawnDate() {
        return despawnDate;
    }

    public void setDespawnDate(Date despawnDate) {
        this.despawnDate = despawnDate;
    }

    public List<String> getPlanets() {
        return planets;
    }

    public void setPlanets(List<String> planets) {
        this.planets = planets;
    }

    public Map<String, Integer> getAttributes() {
        return attributes;
    }

    public void setAttributes(Map<String, Integer> attributes) {
        this.attributes = attributes;
    }

    public String getResourceTypeString() {
        if (resourceTypeString == null && resourceType != null)
            return resourceType.getId();
        return resourceTypeString;
    }

    public void setResourceTypeString(String resourceTypeString) {
        this.resourceTypeString = resourceTypeString;
    }

    public void setAttribute(String attribute, int value) {
        attributes.put(attribute, value);
    }


    @Override
    public String toString() {
        return "GalaxyResource{" +
                "name='" + name + '\'' +
                ", resourceType='" + resourceType + '\'' +
                ", container='" + container + '\'' +
                '}';
    }

    public int getAttribute(String name) {
        return attributes.get(name);
    }
}

File: api/src/main/kotlin/com/lewisjmorgan/harvesterdroid/api/GalaxyResource.kt
package com.lewisjmorgan.harvesterdroid.api

import com.lewisjmorgan.harvesterdroid.api.resource.ResourceType
import java.util.*

class GalaxyResource(var partial: Boolean) {
  constructor(): this(false)

  var name: String = ""
  var container: String = ""

  var spawnDate: Date = Date(0L)
  var despawnDate: Date = Date(0L)

  val planets = ArrayList<String>()
  val attributes = HashMap<String, Int>()

  var resourceType: ResourceType = ResourceType()

  var resourceTypeString: String = ""

  fun getAttribute(name: String): Int {
    return attributes.getOrDefault(name, -1)
  }

  fun setAttribute(attribute: String, value: Int) {
    attributes[attribute] = value
  }

  fun isSpawned(): Boolean {
    return despawnDate != Date(0L)
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as GalaxyResource

    if (partial != other.partial) return false
    if (name != other.name) return false
    if (container != other.container) return false
    if (spawnDate != other.spawnDate) return false
    if (despawnDate != other.despawnDate) return false
    if (planets != other.planets) return false
    if (attributes != other.attributes) return false
//    if (resourceType != other.resourceType) return false
    if (resourceTypeString != other.resourceTypeString) return false

    return true
  }

  override fun hashCode(): Int {
    var result = partial.hashCode()
    result = 31 * result + name.hashCode()
    result = 31 * result + container.hashCode()
    result = 31 * result + spawnDate.hashCode()
    result = 31 * result + despawnDate.hashCode()
    result = 31 * result + planets.hashCode()
    result = 31 * result + attributes.hashCode()
//    result = 31 * result + resourceType.hashCode()
    result = 31 * result + resourceTypeString.hashCode()
    return result
  }

  override fun toString(): String {
    return "GalaxyResource(partial=$partial, name='$name', container='$container', spawnDate=$spawnDate, despawnDate=$despawnDate, planets=$planets, attributes=$attributes, resourceType=$resourceType, resourceTypeString='$resourceTypeString')"
  }
}